strawberryfields/tdm/tdmprogram.py

```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
+                [(sample 4), (sample 6)]],
+            2: [[(sample 1), (sample 3)],
+                [(sample 5), (sample 7)]],
+        }
+
     Args:
         all_samples (dict[int, list]): the raw measured samples
         modes (Sequence[int]): the modes that are measured in the circuit
         N (Sequence[int]): the number of concurrent modes per belt/spatial modes
+        timebins (int): the number of timebins/temporal modes in the program per shot
 
     Returns:
-        new_samples (dict[int, list]): the re-shaped samples
+        dict[int, list]: the re-shaped samples, where each key correspond to a spatial
+            mode and the values have shape ``(shots, timebins)``
     """
+    # calculate the total number of samples and the order in which they were measured
     num_of_values = len([i for j in all_samples.values() for i in j])
-    mode_order = _get_mode_order(num_of_values, N)
+    mode_order = _get_mode_order(num_of_values, N, timebins)
 
-    # go backwards through all_samples and add them into the correct mode
+    # iterate backwards through all_samples and add them into the correct mode
     new_samples = dict()
+    t = 0
     for i, m in enumerate(mode_order):
         idx = modes[i % len(N)]
         if idx not in new_samples:
-            new_samples[idx] = []
-        new_samples[idx].append(all_samples[m].pop(0))
-    return new_samples
+            new_samples[idx] = [[] for _ in range(timebins)]
+        new_samples[idx][t].append(all_samples[m].pop(0)[0])
+
+        # populate each spatial mode in one timebin before moving on to the next timebin
+        # when each timebin has been filled, move to the next shot
+        if (i + 1) % len(N) == 0:
+            t = (t + 1) % timebins
+
+    # transpose each value so that it has shape `(shots, timebins)`
+    return {k: np.array(v).T for k, v in new_samples.items()}
+
+
+def move_vac_modes(samples, N, crop=False):
+    """Post-processing function for TDM samples
+
+    Moves all measured vacuum modes from the first shot of the samples
+    array to the end of the last shot.
+
+    Args:
+        samples (ndarray[float]): samples as received from TDMProgram, with the
+            measured vacuum modes in the first shot
+
+        N (int or Sequence[int]): If an integer, the number of concurrent (or 'alive')
+            modes in each time bin. Alternatively, a sequence of integers
+            may be provided, corresponding to the number of concurrent modes in
+            the possibly multiple bands in the circuit.
+
+    Keyword args:
+        crop (bool): whether to remove all the shots containing measured vacuum
+            modes at the end
+
+    Returns:
+        ndarray[float]: the post-processed samples
+
+    """
+    num_of_vac_modes = np.max(N) - 1
+    shape = samples.shape
+
+    flat_samples = np.ravel(samples)
+    samples = np.append(flat_samples[num_of_vac_modes:], [0] * num_of_vac_modes)
+    samples = samples.reshape(shape)
```

Comment:
Wondering if this could be done without having to ravel/reshape the entire samples array :thinking: Maybe using two indices could work? (There wasn't a good alternative that came to mind though. :confused: )

Reply:
I don't think that this is such a big issue though since (correct me if I'm wrong) I believe `np.ravel` and `.reshape` only changes references and doesn't make any copies if the argument is already a NumPy array. So there's quite little overhead even for large sample arrays. Not sure about the appending though.

Reply:
Hmm, I think you might be right! Checked on an array and there doesn't seem to be considerable overhead added by the ravel/reshape part :+1: